Blog inp frameworks (#7872)
GoogleChrome/web.dev: src/site/content/en/blog/inp-in-frameworks/index.md

   </td>
   <td>Similar to FID for every interaction plus usage of rendering patterns that prioritize key UX updates over other rendering tasks. 
   </td>
  </tr>
   </tbody>
   <caption>FID versus INP: Measurement and optimization</caption>
</table>
</div>

The [Aurora](/introducing-aurora/) team in Chrome works with open-source web frameworks to help developers improve different aspects of the user experience, including performance and CWV metrics. With the introduction of INP, we want to be prepared for the change in CWV metrics for framework-based websites. We have collected data based on the experimental responsiveness metric in CrUX reports. We will share insights and action items to ease the transition to the INP metric for framework-based websites. 

## Experimental responsiveness metric data

An INP below or equal to 200 milliseconds indicates good responsiveness. The CrUX report data and the [CWV Technology Report](https://www.google.com/url?q=https://datastudio.google.com/s/gPxRJmoivLA&sa=D&source=docs&ust=1652475737927418&usg=AOvVaw3ZQ0WOZu2ZGjCuJb23dQYg) for April 2022 give us the following information about responsiveness for popular JavaScript frameworks.

> Note: This measurement includes data from all versions of the frameworks listed. For more mature frameworks, this can include data from versions many years out-of-date.

Reviewer: ```suggestion
{% Aside %}
This measurement includes data from all versions of the frameworks listed. For more mature frameworks, this can include data from versions many years out-of-date.
{% endAside %}
```

<div class="table-wrapper scrollbar">
<table>
<thead>
  <tr>
   <th rowspan="2">Technology
   </th>
   <th colspan="2" align="center">% Passing
   </th>
  </tr>
  <tr>
   <th>% Mobile
   </th>
   <th>Desktop
   </th>
  </tr>
</thead>
  <tbody>
  <tr>
   <td>Angular (v2.0.0+)
   </td>
   <td>19.0
   </td>
   <td>65.5
   </td>
  </tr>
  <tr>
   <td>Next.js
   </td>
   <td>20.2
   </td>
   <td>73.4
   </td>
  </tr>
  <tr>
   <td>Nuxt.js
   </td>
   <td>25.4
   </td>
   <td>84.5
   </td>
  </tr>
  <tr>
   <td>Preact
   </td>
   <td>36.6
   </td>
   <td>90.6
   </td>
  </tr>
  <tr>
   <td>Vue (v2.0.0+)
   </td>
   <td>41.7
   </td>
   <td>90.0
   </td>
  </tr>
  <tr>
   <td>Lit
   </td>
   <td>36.4
   </td>
   <td>75.7
   </td>
  </tr>
  </tbody>
   <caption>CWV technology report - INP data for April 2022</caption>
</table>
</div>

{% Aside 'important' %}

> We recommend not solely making decisions on the frameworks you are choosing based on the numbers above, without being aware of additional nuance. Many different variables contribute to making a framework suitable for your web apps, and the table only reflects INP. Additionally, the used dataset only looks into landing pages which is not the typical use case for some of the listed frameworks. Besides the framework used, several other reasons could affect performance metrics. It is also worth noting that frameworks are often used for different application holotypes, which may play a factor here. The purpose of the data is to indicate the extent of effort required to meet INP goals if your site uses a specific framework.

Reviewer: ```suggestion
We recommend not solely making decisions on the frameworks you are choosing based on the numbers above, without being aware of additional nuance. Many different variables contribute to making a framework suitable for your web apps, and the table only reflects INP. Additionally, the used dataset only looks into landing pages which is not the typical use case for some of the listed frameworks. Besides the framework used, several other factors could affect performance metrics. It is also worth noting that frameworks are often used for different application holotypes, which may play a factor here. The purpose of the data is to indicate the extent of effort required to meet INP goals if your site uses a specific framework.
```

* **React and Next.js:** The [Next.js Script component](/script-component/) helps to address issues caused due to inefficient loading of third-party scripts. [Granular chunking](/granular-chunking-nextjs/) was introduced in Next.js to allow for smaller-sized chunks for shared code. This helps to reduce the amount of unused common code that is downloaded on all pages. We are also working with Next.js to make INP data available as part of their [Analytics](https://nextjs.org/analytics) service.

* **Angular:** Aurora is [partnering with the Angular](https://angular.io/guide/roadmap#explore-hydration-and-server-side-rendering-usability-improvements) team to explore server-side rendering and hydration improvements. We also plan to look into refinements in event handling and change detection to improve INP.

* **Vue and Nuxt.js:** We are exploring avenues for collaboration, mainly in relation to script loading and rendering.

### How are frameworks thinking about improving INP?

> #### React & Next.js

Reviewer: ```suggestion
#### React and Next.js
```

> React.js **time slicing**, implemented through **[startTransition](https://github.com/reactwg/react-18/discussions/41)** and **Suspense**, allows you to opt-in to selective or progressive hydration. This means that hydration isn't a synchronous block. It's done in small slices that are interruptible at any point.

Reviewer: ```suggestion
React.js _time slicing_, implemented through [startTransition](https://github.com/reactwg/react-18/discussions/41) and _Suspense_, allows you to opt-in to selective or progressive hydration. This means that hydration isn't a synchronous block. It's done in small slices that are interruptible at any point.
```

This should help improve INP and enable you to respond more quickly to keystrokes, hover effects during the transition, and clicks. It also helps to keep React apps responsive even for large transitions such as auto-complete. 

> Next.js is working on a **[new routing framework](https://twitter.com/leeerob/status/1521659624516030466)** that will use startTransition by default for route transitions. This goal is to allow Next.js site owners to adopt React time-slicing and improve the responsiveness of route transitions.

Reviewer: ```suggestion
Next.js is working on a [new routing framework](https://twitter.com/leeerob/status/1521659624516030466) that will use startTransition by default for route transitions. This goal is to allow Next.js site owners to adopt React time-slicing and improve the responsiveness of route transitions.
```

#### Angular

The Angular team is exploring several ideas that should also help with INP:

> * **Zoneless**: cuts down on initial bundle size, required code that must load before an app can render anything

Reviewer: ```suggestion
* **Zoneless:** Cuts down on initial bundle size, and required code that must load before an app can render anything.
```

> * **Hydration**: island-style hydration to limit how much of the app needs to be woken up for interaction

Reviewer: ```suggestion
* **Hydration:** Island-style hydration to limit how much of the app needs to be woken up for interaction.
```

> * **Reduce overhead of CD**: make change detection less expensive, find ways to check less of the app, leverage reactive signals about what's changed, etc.

Reviewer: ```suggestion
* **Reduce overhead of CD:** For example, make change detection less expensive, find ways to check less of the app, and leverage reactive signals about what's changed.
```

> * **More granular code-splitting**: make the initial bundle smaller

Reviewer: ```suggestion
* **More granular code-splitting:** Make the initial bundle smaller.
```

> * **Better support for loading indicators**: during SSR re-render, during route navigation, lazy loading operations, etc.

Reviewer: ```suggestion
* **Better support for loading indicators:**: For example, during SSR re-render, during route navigation, and in lazy loading operations.
```

> * **Profiling tools**: better dev tools to understand interaction cost, particularly around change detection cost for specific interactions

Reviewer: ```suggestion
* **Profiling tools:** Better dev tools to understand interaction cost, particularly around change detection cost for specific interactions.
```

---
layout: post
title: How do modern frameworks perform on the new INP metric
subhead: Understand how the new INP metric affects the experience of sites built using JavaScript frameworks and libraries.

> date: 2022-05-14

Reviewer: ```suggestion
date: 2022-05-16
```

* **Rehydration code:** Execution of rehydration code after repainting frames can cause delays. Generating some content statically during build time and routing to this content on events can help to render the content quickly.

From now on, for a good INP score, developers will have to focus on reviewing the code that executes after every interaction on the page and optimize their chunking, rehydration, and loading strategies for both first-party and third-party scripts,  

# How is Aurora addressing INP issues?

Aurora works with frameworks by incorporating best practices to provide baked-in solutions to common problems. We have worked with Next.js, Nuxt.js, Gatsby, and Angular on [solutions](/introducing-aurora/#what-has-our-work-unlocked-so-far) that offer strong defaults within the framework to optimize performance. Following are the highlights of our work in this context:

> * **React/Next.js:** The [Next.js Script component](/script-component/) helps to address issues caused due to inefficient loading of third-party scripts. [Granular chunking](/granular-chunking-nextjs/) was introduced in Next.js to allow for smaller-sized chunks for shared code. This helps to reduce the amount of unused common code that is downloaded on all pages. We are also working with Next.js to make INP data available as part of their [Analytics](https://nextjs.org/analytics) service.

Reviewer: ```suggestion
* **React and Next.js:** The [Next.js Script component](/script-component/) helps to address issues caused due to inefficient loading of third-party scripts. [Granular chunking](/granular-chunking-nextjs/) was introduced in Next.js to allow for smaller-sized chunks for shared code. This helps to reduce the amount of unused common code that is downloaded on all pages. We are also working with Next.js to make INP data available as part of their [Analytics](https://nextjs.org/analytics) service.
```

* <strong>React/Next.js:</strong> The [Next.js Script component](/script-component/) helps to address issues caused due to inefficient loading of third-party scripts. [Granular chunking](/granular-chunking-nextjs/) was introduced in Next.js to allow for smaller-sized chunks for shared code. This helps to reduce the amount of unused common code that is downloaded on all pages. We are also working with Next.js to make INP data available as part of their [Analytics](https://nextjs.org/analytics) service.

* <strong>Angular:</strong> Aurora is [partnering with the Angular](https://angular.io/guide/roadmap#explore-hydration-and-server-side-rendering-usability-improvements) team to explore server-side rendering and hydration improvements. We also plan to look into refinements in event handling and change detection to improve INP.

> * <strong>Vue/Nuxt.js:</strong> We are exploring avenues for collaboration, mainly in relation to script loading and rendering.

Reviewer: ```suggestion
* **Vue and Nuxt.js:** We are exploring avenues for collaboration, mainly in relation to script loading and rendering.
```
